### PR TITLE
Run dartfix to upgrade code to use int to double inference

### PR DESCRIPTION
**Pull request by mit-mit**
I ran the latest version of https://pub.dartlang.org/packages/dartfix to make sure the Gallery uses the new Dart 2.1 int to double inference feature

### CODE REVIEW

**Review by DanTup — DISMISSED** [review]
(RS?)LGTM - I scanned through the changes and didn't see anything that wasn't exactly just removing a `.0` on a double so if this feature works as advertised, looks good!

File: examples/flutter_gallery/lib/demo/animation/home.dart [hunk 1]
```diff
@@ -341,7 +341,7 @@ class _AllSectionsView extends AnimatedWidget {
 
     return CustomMultiChildLayout(
       delegate: _AllSectionsLayout(
-        translation: Alignment((selectedIndex.value - sectionIndex) * 2.0 - 1.0, -1.0),
+        translation: Alignment((selectedIndex.value - sectionIndex) * 2.0 - 1.0, -1),
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/animation/home.dart]: And the others too

File: examples/flutter_gallery/lib/demo/material/backdrop_demo.dart [hunk 2]
```diff
@@ -301,9 +301,9 @@ class _BackdropDemoState extends State<BackdropDemo> with SingleTickerProviderSt
 
     final double flingVelocity = details.velocity.pixelsPerSecond.dy / _backdropHeight;
     if (flingVelocity < 0.0)
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/backdrop_demo.dart]: Here too?

File: examples/flutter_gallery/lib/demo/material/backdrop_demo.dart [hunk 3]
```diff
@@ -301,9 +301,9 @@ class _BackdropDemoState extends State<BackdropDemo> with SingleTickerProviderSt
...
-      _controller.fling(velocity: math.max(2.0, -flingVelocity));
+      _controller.fling(velocity: math.max(2, -flingVelocity));
     else if (flingVelocity > 0.0)
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/backdrop_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/material/backdrop_demo.dart [hunk 4]
```diff
@@ -301,9 +301,9 @@ class _BackdropDemoState extends State<BackdropDemo> with SingleTickerProviderSt
...
-      _controller.fling(velocity: math.min(-2.0, -flingVelocity));
+      _controller.fling(velocity: math.min(-2, -flingVelocity));
     else
       _controller.fling(velocity: _controller.value < 0.5 ? -2.0 : 2.0);
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/backdrop_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart [hunk 5]
```diff
@@ -78,7 +78,7 @@ class CustomIcon extends StatelessWidget {
   Widget build(BuildContext context) {
     final IconThemeData iconTheme = IconTheme.of(context);
     return Container(
-      margin: const EdgeInsets.all(4.0),
+      margin: const EdgeInsets.all(4),
       width: iconTheme.size - 8.0,
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart [hunk 6]
```diff
@@ -78,7 +78,7 @@ class CustomIcon extends StatelessWidget {
...
       height: iconTheme.size - 8.0,
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart [hunk 7]
```diff
@@ -91,11 +91,11 @@ class CustomInactiveIcon extends StatelessWidget {
   Widget build(BuildContext context) {
     final IconThemeData iconTheme = IconTheme.of(context);
     return Container(
-      margin: const EdgeInsets.all(4.0),
+      margin: const EdgeInsets.all(4),
       width: iconTheme.size - 8.0,
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart [hunk 8]
```diff
@@ -91,11 +91,11 @@ class CustomInactiveIcon extends StatelessWidget {
...
       height: iconTheme.size - 8.0,
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/bottom_navigation_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/material/elevation_demo.dart [hunk 9]
```diff
@@ -14,25 +14,25 @@ class _ElevationDemoState extends State<ElevationDemo> {
...
+      16,
+      24,
     ];
 
     return elevations.map<Widget>((double elevation) {
       return Center(
         child: Card(
-          margin: const EdgeInsets.all(20.0),
+          margin: const EdgeInsets.all(20),
           elevation: _showElevation ? elevation : 0.0,
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/elevation_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/material/grid_list_demo.dart [hunk 10]
```diff
@@ -370,9 +370,9 @@ class GridListDemoState extends State<GridListDemo> {
               bottom: false,
               child: GridView.count(
                 crossAxisCount: (orientation == Orientation.portrait) ? 2 : 3,
-                mainAxisSpacing: 4.0,
-                crossAxisSpacing: 4.0,
-                padding: const EdgeInsets.all(4.0),
+                mainAxisSpacing: 4,
+                crossAxisSpacing: 4,
+                padding: const EdgeInsets.all(4),
                 childAspectRatio: (orientation == Orientation.portrait) ? 1.0 : 1.3,
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/material/grid_list_demo.dart]: And here?

File: examples/flutter_gallery/lib/demo/shrine/shrine_page.dart [hunk 11]
```diff
@@ -37,7 +37,7 @@ class ShrinePage extends StatefulWidget {
 
 /// Defines the Scaffold, AppBar, etc that the demo pages have in common.
 class ShrinePageState extends State<ShrinePage> {
-  double _appBarElevation = 0.0;
+  double _appBarElevation = 0;
 
   bool _handleScrollNotification(ScrollNotification notification) {
     final double elevation = notification.metrics.extentBefore <= 0.0 ? 0.0 : 1.0;
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/demo/shrine/shrine_page.dart]: And here?

File: examples/flutter_gallery/lib/gallery/backdrop.dart [hunk 12]
```diff
@@ -236,9 +236,9 @@ class _BackdropState extends State<Backdrop> with SingleTickerProviderStateMixin
 
     final double flingVelocity = details.velocity.pixelsPerSecond.dy / _backdropHeight;
     if (flingVelocity < 0.0)
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/gallery/backdrop.dart]: And here?

File: examples/flutter_gallery/lib/gallery/backdrop.dart [hunk 13]
```diff
@@ -236,9 +236,9 @@ class _BackdropState extends State<Backdrop> with SingleTickerProviderStateMixin
...
-      _controller.fling(velocity: math.max(2.0, -flingVelocity));
+      _controller.fling(velocity: math.max(2, -flingVelocity));
     else if (flingVelocity > 0.0)
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/gallery/backdrop.dart]: And here?

File: examples/flutter_gallery/lib/gallery/backdrop.dart [hunk 14]
```diff
@@ -236,9 +236,9 @@ class _BackdropState extends State<Backdrop> with SingleTickerProviderStateMixin
...
-      _controller.fling(velocity: math.min(-2.0, -flingVelocity));
+      _controller.fling(velocity: math.min(-2, -flingVelocity));
     else
       _controller.fling(velocity: _controller.value < 0.5 ? -2.0 : 2.0);
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/gallery/backdrop.dart]: And here?

File: examples/flutter_gallery/lib/gallery/home.dart [hunk 15]
```diff
@@ -325,8 +325,8 @@ class _GalleryHomeState extends State<GalleryHome> with SingleTickerProviderStat
     final MediaQueryData media = MediaQuery.of(context);
     final bool centerHome = media.orientation == Orientation.portrait && media.size.height < 800.0;
```
Comment (tvolkert) [inline examples/flutter_gallery/lib/gallery/home.dart]: And here?

File: examples/flutter_gallery/test/demo/material/drawer_demo_test.dart [hunk 16]
```diff
@@ -21,7 +21,7 @@ void main() {
     final double initialTopItemSaneY = tester.getTopLeft(find.text('Drawer item A')).dy;
     expect(initialTopItemSaneY, greaterThan(0.0));
 
-    await tester.drag(find.text('Drawer item B'), const Offset(0.0, 400.0));
+    await tester.drag(find.text('Drawer item B'), const Offset(0, 400));
     await tester.pump();
 
     expect(tester.getTopLeft(find.byType(UserAccountsDrawerHeader)).dy, 0.0);
```
Comment (tvolkert) [inline examples/flutter_gallery/test/demo/material/drawer_demo_test.dart]: And here?

**Comment by tvolkert** [comment]
There are two meta-questions to go along with my comments:

1. Will be implicit conversion work in those cases (cases like `doubleValue + 8.0` being converted to `doubleValue + 8`, or `doubleValue <= 8.0` being converted to `doubleValue <= 8`)?  If not, then I think this implicit double conversion feature is going to cause more harm than good, and we need to fix that ASAP.
2. If the cases I pointed out _are_ supported, then presumably they were missed in this PR because the linter didn't catch those cases?  If that's true, then we should see why the linter didn't catch them.

**Comment by mit-mit** [comment]
cc @danrubel, it looks like there are a number of additional places where we could change doubles to ints?

**Comment by lrhn** [comment]
The *integer-double-literal* feature only means that you can change a double literal to an integer-double literal when it occurs in a double context, because then it will still evaluate to a double value, and hence the change is semantics preserving.

There is no double context for the second operand of `doubleValue + 8.0`. You can use an integer there (e.g., a variable with type `int`). So, if we change `8.0` to `8`, we change the program semantics. It will still give the same result, but that's just because `double.operator+` starts by doing `toDouble` on its argument. It's not the same program with the same *semantics*, just one with the same *result*.

We can definitely do program transformations that only preserve the result (and visible side effects), like changing double literals as second-operands to double operators into integer literals. We could have done that before the integer-double-literal feature too, it's a completely unrelated change (it might just be more stylistically consistent to do it now).
We just have to be very, very careful because we *are* changing program semantics.

For example:
```dart
double d = 1;
d = d + 18446744073709552000.0;
```
If we change `18446744073709552000.0` to `18446744073709552000`, it's becomes a compile-time error. Since it's not in a double context, we only allow valid integer literal values, and 2^64 is not a valid *integer* literals. In a double context, it would be interpreted as a double value and work.
(Again a sign that this is unrelated to integer-double-literals).

File: examples/flutter_gallery/lib/demo/material/backdrop_demo.dart (hunk 4 above)
Comment (lrhn) [reply examples/flutter_gallery/lib/demo/material/backdrop_demo.dart]: Here I hope we can eventually push the double context into the branches. It's very annoying that we don't.

File: examples/flutter_gallery/lib/demo/material/backdrop_demo.dart (hunk 3 above)
Comment (lrhn) [reply examples/flutter_gallery/lib/demo/material/backdrop_demo.dart]: There is absolutely no way we can convert 2.0 here to an integer, not without explicitly writing the `<double>` type argument. There is no double context here except the one inferred from the types of the argument, so we can't use the context to decide the types of the arguments.
If we just make it an integer, then the inferred type argument of `math.max` becomes `<num>`, and the result might be the integer 2.

File: examples/flutter_gallery/test/demo/material/drawer_demo_test.dart (hunk 16 above)
Comment (lrhn) [reply examples/flutter_gallery/test/demo/material/drawer_demo_test.dart]: There is no context type here, the second argument of `expect` has type `dynamic`. There is no way to guess whether changing 0.0 to 0 is correct, short of recognizing `package:test`'s `expect` function and knowing how it uses its second argument if it is a number value. (It uses it as the first operand of an `==` operator, if it had been the second operand, the change would just be unsafe).

File: examples/flutter_gallery/lib/demo/material/backdrop_demo.dart (hunk 3 above)
Comment (DanTup) [reply examples/flutter_gallery/lib/demo/material/backdrop_demo.dart]: I should leave the reviews to the experts 😄  So is this a bug in `dartfix`?

I think this illustrates the issue:

```
  final a = math.max(2.0, 1.5); // analysis reports a as 'double'
  final b = math.max(2, 1.5); // analysis reports b as 'num'

  double a2 = a;
  double b2 = b; // runtime error: 'int' is not a subtype of type 'double'
```

**Comment by tvolkert** [comment]
Each of those cases makes sense when you go through it.

However, I'm worried that the cognitive overhead right now of knowing when you're in a double context and when you're not (and the implications of using an integer in the wrong place) is going to bite developers in subtle ways, and they'll either (a) blame the language, or (b) just always use explicit doubles, which defeats the purpose of the feature. :-/

**Comment by mit-mit** [comment]
cc @InMatrix @leafpetersen any thoughts on this?

**Pull request closed without merging**